Initial release (#1055)
bitcoin-core/secp256k1: doc/release-process.md

# Release Process

1. Open PR to master that

>    * adds release notes to `doc/CHANGELOG.md` (set the date of the new release to dummy date `YYYY-MM-DD`), and

Reviewer: nit: new path of CHANGELOG.md (more instances below)

>    * if this is **not** a patch release, updates `_PKG_VERSION_{MAJOR,MINOR}` and `_LIB_VERSIONS_*` in `configure.ac`.

Reviewer 2: ```suggestion
   * if this is **not** a patch release, updates `_PKG_VERSION_{MAJOR,MINOR}` and `_LIB_VERSION_*` in `configure.ac`.
```

2. After the PR is merged,
   * if this is **not** a patch release, create a release branch with name `MAJOR.MINOR`.
     Make sure that the branch contains the right commits.
     Create commit on the release branch that sets `_PKG_VERSION_IS_RELEASE` in `configure.ac` to `true` and replaces the dummy date in `doc/CHANGELOG.md` with the current date.
   * if this **is** a patch release open a pull request that
     * includes the bugfixes to the `MAJOR.MINOR` branch,
     * includes the release note commit from master,
     * replaces the dummy date in `doc/CHANGELOG.md` with the current date,

>      * bumps of `_PKG_VERSION_BUILD` and `_LIB_VERSIONS_*` in `configure.ac`.

Reviewer 2: ```suggestion
     * bumps of `_PKG_VERSION_BUILD` and `_LIB_VERSION_*` in `configure.ac`.
```

This document outlines the process for releasing versions of the form `$MAJOR.$MINOR.$PATCH`.

We distinguish between two types of releases: *regular* and *maintenance* releases.
Regular releases are releases of a new major or minor version as well as patches of the most recent release.
Maintenance releases, on the other hand, are required for patches of older releases.

You should coordinate with the other maintainers on the release date, if possible.
This date will be part of the release entry in [CHANGELOG.md](../CHANGELOG.md) and it should match the dates of the remaining steps in the release process (including the date of the tag and the GitHub release).
It is best if the maintainers are present during the release, so they can help ensure that the process is followed correctly and, in the case of a regular release, they are aware that they should not modify the master branch between merging the PR in step 1 and the PR in step 3.

This process also assumes that there will be no minor releases for old major releases.

## Regular release

1. Open a PR to the master branch with a commit (using message `"release: prepare for $MAJOR.$MINOR.$PATCH"`, for example) that

>        * adds release notes to [CHANGELOG.md](../CHANGELOG.md) (make sure to include an entry for `### ABI Compatibility`) and

Author: Fixed.

       * updates `_PKG_VERSION_*`, `_LIB_VERSION_*`, and sets `_PKG_VERSION_IS_RELEASE` to `true` in `configure.ac`.
2. After the PR is merged, tag the commit and push it:
   ```
   RELEASE_COMMIT=<merge commit of step 1>
   git tag -s v$MAJOR.$MINOR.$PATCH -m "libsecp256k1 $MAJOR.$MINOR.$PATCH" $RELEASE_COMMIT
   git push git@github.com:bitcoin-core/secp256k1.git v$MAJOR.$MINOR.$PATCH
   ```
3. Open a PR to the master branch with a commit (using message `"release: bump version after $MAJOR.$MINOR.$PATCH"`, for example) that sets `_PKG_VERSION_IS_RELEASE` to `false` and `_PKG_VERSION_PATCH` to `$PATCH + 1` and increases `_LIB_VERSION_REVISION`.
  * If other maintainers are not present to approve the PR, it can be merged without ACKs.
4. Create a new GitHub release with a link to the corresponding entry in [CHANGELOG.md](../CHANGELOG.md).

## Maintenance release

Note that bugfixes only need to be backported to releases for which no compatible release without the bug exists.

> 1. Open PR to the master branch that includes a commit (with commit message `"release: prepare for $MAJOR.$MINOR.$PATCH"`, for example) that adds release notes to [CHANGELOG.md](../CHANGELOG.md).

Reviewer: > I'd say the only required step is the creation of a `3.1` branch, branching off the `v3.1.0` merge commit, and adding the release notes relevant for `v3.1.1` there.

Hm yeah, I don't know. So far I held the belief that it's a good thing to treat the changelog more like a "global" thing that lists all releases (regular and maintenance across branches). Because it makes sense to keep a list of all releases somehwere. And then the canonical place for it is master. (Or outside the repo, e.g., in a wiki, but that comes with other difficulties.) 

How do other projects handle this?

(Happy to postpone this and all other things related to  maintenance releases to the future)

> 1. Open PR to the master branch that includes a commit (with commit message `"release: prepare for $MAJOR.$MINOR.$PATCH"`, for example) that adds release notes to [CHANGELOG.md](../CHANGELOG.md).

Author: > How about moving step (1) to be after step (6), and calling it Add $MAJOR.$MINOR.$PATCH to release notes? Step (1) happens in a different branch than everything else, so the order shouldn't matter.

sgtm

> 1. Open PR to the master branch that includes a commit (with commit message `"release: prepare for $MAJOR.$MINOR.$PATCH"`, for example) that adds release notes to [CHANGELOG.md](../CHANGELOG.md).

Author: done

# Release Process

This document outlines the process for releasing versions of the form `$MAJOR.$MINOR.$PATCH`.

We distinguish between two types of releases: *regular* and *maintenance* releases.
Regular releases are releases of a new major or minor version as well as patches of the most recent release.
Maintenance releases, on the other hand, are required for patches of older releases.

> Note that adding a release entry to [CHANGELOG.md](../CHANGELOG.md) requires a release date. This date should match the dates of the remaining steps in the release process, including the date of the tag and the GitHub release.

Reviewer 2: Is ti worth mentioning in the `CHANGELOG.md` for each regular release whether it is ABI backward compatible or not?
A clearl statement about ABI backward compatibility will also help to modify `_LIB_VERSION_*` variables properly.

Author: It would make more sense to me to do the opposite and mention that we make little guarantees about the library version. We don't have methods to accurately determine ABI compatibility and so we can either put reasonable effort into attempting proper library versioning or not having library versions at all. So far the consensus seemed to be that we do the former.

Author: OTOH if our resonable-effort approach to library versioning would indicate that it's not consistent with semantic versioning, then I agree that it makes sense to mention that in the changelog. You think that's worth adding to the `release-process.md`?

Author: > I don't think it's unreasonable to at least try and have a very good chance of getting it right

Good, then there doesn't seem to be disagreement.

If we want to mention ABI compatibility in every  release note, we could add a section `### ABI compatibility` (or so) to the release notes that (somewhat cautiously) mentions how we set the library version.

Author: I added a section `### ABI compatibility` to the changelog and mention it in the release process.

This process also assumes that there will be no minor releases for old major releases.

## Regular release

> 1. Open PR to the master branch that includes the following commits:

Reviewer: Urghs, this is a good point. (And this was not a problem with the previous approach...)

> Don't care about it

 Well we anyway need to open the PR for the right date to get the release date right... So yeah, this is doable but it still seems error-prone. (And releases are the one thing where we don't want to make mistakes, because we cannot correct them without another release.)

> Use two separate PRs, and tag the merge-commit of the first.

Ok, that works. I tried to avoid two PRs but ok.  What if we have a single PR, merge that locally with the merge tool, tag the merge, let the merging maintainer add the "bump after release" commit manually, and then push the result. That saves the second PR and having that one commit with PR is harmless. (And if there's a problem with the "bump after release" commit, then we can just fix it on master and it won't affect the release.) 

> Use two commits (or at least the first one) that's not part of a PR, but just immediately pushed by a maintainer to the master branch.

Is that what I just wrote above? I'm not sure. You say at least the first commit should be manual and not part of the PR, and I don't see how that would work. (In my suggestion, the *second* commit is done manually.)

> 1. Open PR to the master branch that includes the following commits:

Reviewer: > I don't see why the second commit matters? It's past the tag, so it can't affect it.
> 
> All that matters is that there is a stable point on the master branch after the first commit. That can be achieved by it being in a separate PR, and using the merge commit of that PR as the tag, or by just pushing a commit and marking that as the tag.

Yes, we're in full agreement here! 

So I think we should
1. tag the merge commit of a PR
2. then get the "release: update master ..." commit on master (with or without a PR). 

Now that I think about  this again, I tend so say let's make a PR so that we never break the implicit "don't push without PR" rule (this also means other maintainers will see it etc.) But to make sure the process is not overly complicated and interactive, I suggest that the maintainer preparing the release simply merges this second PR without an ACK. (And indeed, even if we then spot a mistake in the second PR, it won't affect the release).

> 1. Open PR to the master branch that includes the following commits:

Author: That's quite unfortunate, I did not consider this.

> So I think we should
> 1. tag the merge commit of a PR
> 2. then get the "release: update master ..." commit on master (with or without a PR).

One issue with this approach is that `IS_RELEASE` is set to true in the merge commit as well as in the commit that set it to true. Therefore, we have two (maybe significantly) different states of the repo that are marked as `IS_RELEASE`. Another problem with this approach is that the repo is locked after the first PR until the second commit lands in the repo (unless you ignore `IS_RELEASE` and `VERSION_PATCH` for a while).

Another approach is that we revert the release process back to a version where we set the patch version of master to `99` and branch off from before we set `IS_RELEASE` which is preserved [here](https://github.com/jonasnick/secp256k1/blob/initial-release-backup2/doc/release-process.md). This version, however, has some subtle edge cases that made me prefer the current version of the release process.

I would have no problem with the approach that locks the repo until the PR is merged ("don't care about it") if we could reasonably ensure that no mistakes will happen. But I have no better ideas than mentioning it in the release process and in the OP of the PR.

> 1. Open PR to the master branch that includes the following commits:

Author: As per our discussion elsewhere I changed the release process to @real-or-random's suggestion above.

# Release Process

This document outlines the process for releasing versions of the form `$MAJOR.$MINOR.$PATCH`.

We distinguish between two types of releases: *regular* and *maintenance* releases.
Regular releases are releases of a new major or minor version as well as patches of the most recent release.
Maintenance releases, on the other hand, are required for patches of older releases.

> Note that adding a release entry to CHANGELOG.md requires a release date. This date should match the dates of the remaining steps in the release process, including the date of the tag and the GitHub release.

Reviewer: ```suggestion
Note that adding a release entry to [CHANGELOG.md](../CHANGELOG.md) requires a release date. This date should match the dates of the remaining steps in the release process, including the date of the tag and the GitHub release.
```

Author: fixed

This process also assumes that there will be no minor releases for old major releases.

## Regular release

1. Open PR to the master branch that includes the following commits:
   1. A commit (with commit message `"release: prepare for $MAJOR.$MINOR.$PATCH"`, for example) that
       * adds release notes to [CHANGELOG.md](../CHANGELOG.md) and
       * updates `_PKG_VERSION_*`, `_LIB_VERSION_*`, and sets `_PKG_VERSION_IS_RELEASE` to `true` in `configure.ac`.

>    2. A commit (with commit message `"release: update master after $MAJOR.$MINOR.$PATCH"`, for example) that sets `_PKG_VERSION_IS_RELEASE` to `false` and `_PKG_VERSION_PATCH` to `$PATCH + 1` and increases `_LIB_VERSION_REVISION`.

Reviewer: ```suggestion
   2. A commit (with commit message `"release: Bump version after $MAJOR.$MINOR.$PATCH"`, for example) that sets `_PKG_VERSION_IS_RELEASE` to `false` and `_PKG_VERSION_PATCH` to `$PATCH + 1` and increases `_LIB_VERSION_REVISION`.
```
Mentioning "master" feels somewhat wrong because the commit will end up on other branches, too. I think this is minimally clearer. 

(My suggestion includes starting with a capital letter after the colon in the commit messages, because I find this more readable. But this is just my preference, feel free to stick to lower case.)

Author: true, fixed (and sticked with lowercase for now)

2. After the PR is merged, tag the commit:
   ```

>    BRANCH_COMMIT=<commit of step 1.1>

Reviewer: ```suggestion
   BRANCH_COMMIT=<commit of step 1.i> 
```
That's how it's rendered.

Author: fixed

   git tag -s v$MAJOR.$MINOR.$PATCH -m "libsecp256k1 $MAJOR.$MINOR.$PATCH" $BRANCH_COMMIT
   ```
3. Push tag:
   ```
   git push git@github.com:bitcoin-core/secp256k1.git v$MAJOR.$MINOR.$PATCH
   ```
4. Create a new GitHub release with a link to the corresponding entry in [CHANGELOG.md](../CHANGELOG.md).

## Maintenance release

> Note that some releases do not require maintenance, even if they include bugs.

Author: done

> This is the case when users of the library are able to update to a newer, compatible release that does not include the bug.

Reviewer: ```suggestion
This is the case when users of the library are able to update to a newer, backwards-compatible release that does not include the bug.
```

Author: Chose sipa's suggestion below.

1. Open PR to the master branch that includes a commit (with commit message `"release: prepare for $MAJOR.$MINOR.$PATCH"`, for example) that adds release notes to [CHANGELOG.md](../CHANGELOG.md).
2. If `$PATCH = 1`, create maintenance branch `$MAJOR.$MINOR`:
     ```

>      BRANCH_COMMIT=<id of commit that set _PKG_VERSION_* to $MAJOR.$MINOR>

Reviewer: Couldn't we just use the tag here?
```suggestion
     BRANCH_COMMIT=v$MAJOR.$MINOR.$PATCH
```
If yes, then maybe it fits on a single line. Not sure if that's more readable.

Author: done (should be `v$MAJOR.$MINOR.0`)
